routing/router.go

```diff
@@ -1728,12 +1729,47 @@ func (r *ChannelRouter) preparePayment(payment *LightningPayment) (
 func (r *ChannelRouter) SendToRoute(hash lntypes.Hash, route *route.Route) (
 	lntypes.Preimage, error) {
 
+	return r.SendToRouteMPP(hash, route, 0, [32]byte{})
+}
+
+// SendToRouteMPP attempts to send a payment with the given hash through the
+// provided route. This function is blocking and will return the obtained
+// preimage if the payment is successful or the full error in case of a failure.
+// Providing a total > 0 and a non-zero addr will cause this method to send an
+// MPP shard with those parameters passed via the option_mpp tlv record to the
+// receiver. Otherwise, this call will send a regular single shot payment.
+func (r *ChannelRouter) SendToRouteMPP(hash lntypes.Hash, route *route.Route,
+	total lnwire.MilliSatoshi, addr lntypes.Hash) (lntypes.Preimage, error) {
+
 	// Create a payment session for just this route.
 	paySession := r.cfg.SessionSource.NewPaymentSessionForRoute(route)
 
 	// Calculate amount paid to receiver.
 	amt := route.TotalAmount - route.TotalFees()
 
+	useMPP := total > 0
+
+	// Verify mpp params, if applicable.
+	switch {
+
+	// The amount of the shard should not be greater than the total.
+	case useMPP && amt > total:
+		return [32]byte{}, fmt.Errorf("amt %v cannot be greater "+
+			"than total %v", amt, total)
+
+	// The payment address should not be empty.
+	case useMPP && addr == lntypes.ZeroHash:
+		return [32]byte{}, fmt.Errorf("payment_addr must be populated " +
+			"for mpp")
+	}
+
+	// If we're using mpp, set the amount in the payment creation info to
+	// the total amount. The route will contain the amount paid in the
+	// shard.
+	if useMPP {
+		amt = total
+	}
+
```

Comment:
All of these changes check `useMPP` first. It could all happen in one `if` clause.